src/charm.py

```diff
@@ -193,6 +208,36 @@ def _update_fiveg_n3_relation_data(self) -> None:
                 upf_ip_address=upf_access_ip_address,
             )
 
+    def _update_fiveg_n4_relation_data(self) -> None:
+        """Publishes UPF hostname and the N4 port in the `fiveg_n4` relation data bag."""
+        fiveg_n4_relations = self.model.relations.get("fiveg_n4")
+        if not fiveg_n4_relations:
+            logger.info("No `fiveg_n4` relations found.")
+            return
+        for fiveg_n4_relation in fiveg_n4_relations:
+            self.fiveg_n4_provider.publish_upf_n4_information(
+                relation_id=fiveg_n4_relation.id,
+                upf_hostname=self._get_n4_upf_hostname(),
+                upf_n4_port=PFCP_PORT,
+            )
+
+    def _get_n4_upf_hostname(self) -> str:
+        """Returns the UPF hostname to be exposed over the `fiveg_n4` relation.
+
+        If a configuration is provided, it is returned. If that is
+        not available, returns the hostname of the external LoadBalancer
+        Service. If the LoadBalancer Service does not have a hostname,
+        returns the internal Kubernetes service FQDN.
+
+        Returns:
+            str: Hostname of the UPF
+        """
+        if configured_hostname := self._get_external_upf_hostname_config():
+            return configured_hostname
+        elif lb_hostname := self._upf_external_service_hostname():
+            return lb_hostname
+        return self._upf_external_hostname
```

Comment:
in the config above it says `it will default to the internal Kubernetes FQDN of the service.`, I'm a bit confused as this here say `external_hostname`

Reply:
Yes, this is unfortunate. It's actually the internal hostname of the external service :rofl: 
Will change it.

Reply:
Done.